osgar/test_zmqrouter.py

```diff
@@ -0,0 +1,252 @@
+import ast
+import datetime
+import logging
+import os
+import pathlib
+import shutil
+import tempfile
+import time
+import unittest
+
+from unittest.mock import MagicMock
+
+from threading import Thread
+
+import osgar.logger
+
+from osgar.zmqrouter import _Router, _Bus, record
+
+
+class Noop:
+    def __init__(self, config, bus):
+        self.bus = bus
+        self.bus.register("output")
+
+    def start(self):
+        pass
+
+    def join(self, timeout=None):
+        pass
+
+
+class Publisher:
+    def __init__(self, config, bus):
+        self.bus = bus
+        self.output_name = config["output"].split(":")[0] # drop any possible suffix
+        self.bus.register(config["output"])
+
+    def start(self):
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    def run(self):
+        count = 10
+        for i in range(count):
+            dt = self.bus.publish(self.output_name, i)
+            time.sleep(0.01)
+            #print("  published", i, dt)
+
+
+class PublisherListener:
+    def __init__(self, config, bus):
+        self.bus = bus
+        self.output_name = config["output"].split(":")[0] # drop any possible suffix
+        self.bus.register(config["output"])
+
+    def start(self):
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    def run(self):
+        count = 10
+        for i in range(count):
+            dt, channel, value = self.bus.listen()
+            assert value == i
+            time.sleep(0.15)
+            dt = self.bus.publish(self.output_name, i)
+            #print(self.bus.name, dt, channel, value)
+
+class NoQuit:
+    def __init__(self, config, bus):
+        self.bus = bus
+        self.bus.register()
+
+    def start(self):
+        self.thread = Thread(target=self.run)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    def run(self):
+        while True:
+            time.sleep(1)
+
+
+class Test(unittest.TestCase):
```

Comment:
I would use name `ZMQRouterTest` so it is easier to distinguish the class

Reply:
I have picked the generic name intentionally. When I work on tests, I have an editor opened. The editor shows me the name of the method I am working on and the name of the file. The name of the class is somewhere off the screen. But when I want to run a single test, I need to know the name of the class. Having a generic name there, I don't have to go look it up (one less thing to worry about). The class serves no real purpose, IMHO it is quite superficial - pytest actually manages to work without such a class (which would be even better but there is no such choice for unittest IIRC).